src/client.rs

```diff
@@ -28,17 +28,50 @@ macro_rules! quote {
     };
 }
 
+trait OptionExt<E> {
+    fn err(self) -> std::result::Result<(), E>;
+}
+
+impl<E> OptionExt<E> for Option<E> {
+    fn err(self) -> std::result::Result<(), E> {
+        match self {
+            Some(e) => Err(e),
+            None => Ok(()),
+        }
+    }
+}
+
 fn validate_str(value: &str) -> Result<String> {
     let quoted = quote!(value);
-    if quoted.find('\n').is_some() {
-        return Err(Error::Validate(ValidateError('\n')));
-    }
-    if quoted.find('\r').is_some() {
-        return Err(Error::Validate(ValidateError('\r')));
-    }
+    quoted
+        .matches(|c| c=='\n' || c=='\r')
+        .next()
+        .and_then(|s| s.chars().next())
+        .map(|offender| Error::Validate(ValidateError(offender)))
+        .err()?;
     Ok(quoted)
 }
 
+fn validate_str_noquote(value: &str) -> Result<&str> {
+    value
+        .matches(|c| c=='\n' || c=='\r')
+        .next()
+        .and_then(|s| s.chars().next())
+        .map(|offender| Error::Validate(ValidateError(offender)))
+        .err()?;
+    Ok(value)
+}
+
+fn validate_sequence_set(value: &str) -> Result<&str> {
+    value
+        .matches(|c: char| c.is_ascii_whitespace())
```

Comment:
Is it documented anywhere that a sequence set cannot contain whitespace?

Reply:
I believe so. The formal grammar at https://tools.ietf.org/html/rfc3501#page-90 defines a sequence set in ABNF as follows:

```
sequence-set = (seq-number / seq-range) *("," sequence-set)
seq-range = seq-number ":" seq-number
seq-number = nz-number / "*"
nz-number       = digit-nz *DIGIT
digit-nz        = %x31-39
```

Note how sequence-set, and the rules reachable from sequence-set, only contain the terminals DIGIT, and "*", ":", ","; and not the terminals SP, LWSP or similar.

Reply:
Would you mind adding a link to that in the code, just for others who may read the code years from now?

Reply:
And while you're at it, could you also add a comment as to why we don't quote the fetch query argument?

Reply:
I've added doc-comments on all three of these functions.

Aside: reading section 9 of the IMAP RFC gives me the idea that a not-so-stringly-typed API could consist of builders that correspond to the (relevant and nontrivial parts of the) grammar contained therein.

Reply:
Beautiful, thank you!

And yes, I agree, it seems pretty reasonable to construct the non-stringly-typed API using the building blocks from the spec. I wonder if we can leverage `imap-proto` to a greater extent there?

Reply:
Funny you should mention it, this very thing appears on the TODO-list in imap-proto's repo.